apis/python/src/tiledb/vector_search/ingestion.py

```diff
@@ -1082,12 +1082,17 @@ def update_centroids():
                 new_centroid_thread_sums.append(new_centroid_sums_queue)
                 new_centroid_counts_queue = mp.Queue()
                 new_centroid_thread_counts.append(new_centroid_counts_queue)
+
+                start = i
+                end = i + batch_size
+                if end > len(vectors):
+                    end = len(vectors)
                 worker = mp.Process(
                     target=generate_new_centroid_per_thread,
                     args=(
                         thread_id,
-                        i,
-                        i + batch_size,
+                        start,
+                        end,
                         new_centroid_sums_queue,
                         new_centroid_counts_queue,
```

Comment:
We make this change because of a crash I got when running this code:
```python
def test_cloud_ivf_flat_random_sampling(self):
    # NOTE(paris): This was also tested with the following (and also with mode=Mode.BATCH):
    # source_uri = "tiledb://TileDB-Inc/ann_sift1b_raw_vectors_col_major"
    # training_sample_size = 1000123
    source_uri = "tiledb://TileDB-Inc/sift_10k"
    queries_uri = "test/data/siftsmall/siftsmall_query.fvecs"
    gt_uri = "test/data/siftsmall/siftsmall_groundtruth.ivecs"
    index_uri = CloudTests.ivf_flat_random_sampling_index_uri
    k = 100
    nqueries = 100
    nprobe = 20
    max_sampling_tasks = 13
    training_sample_size = 1234

    queries = load_fvecs(queries_uri) 
    gt_i, gt_d = get_groundtruth_ivec(gt_uri, k=k, nqueries=nqueries)
    
    index = vs.ingest(
        index_type="IVF_FLAT",
        index_uri=index_uri,
        source_uri=source_uri,
        training_sampling_policy=vs.ingestion.TrainingSamplingPolicy.RANDOM,
        training_sample_size=training_sample_size,
        max_sampling_tasks=max_sampling_tasks,
        config=tiledb.cloud.Config().dict(),
        # input_vectors_per_work_item=1000123,
        mode=Mode.BATCH,
        verbose=True,
    )

    check_training_input_vectors(
        index_uri=index_uri, 
        expected_training_sample_size=training_sample_size, 
        expected_dimensions=queries.shape[1])

    _, result_i = index.query(queries, k=k, nprobe=nprobe)
    assert accuracy(result_i, gt_i) > MINIMUM_ACCURACY
```

The error is here https://cloud.tiledb.com/activity/tasks/f033eada-1e59-451f-bf8e-6f1393202267, and the log:
<details>
<summary>Click to view log</summary>
```
(TileDB-Vector-Search-3) ~/repo/TileDB-Vector-Search-3/apis/python pytest test/test_cloud.py -s                  
================================== test session starts ===================================
platform darwin -- Python 3.9.18, pytest-7.4.3, pluggy-1.3.0
rootdir: /Users/parismorgan/repo/TileDB-Vector-Search-3/apis/python
collected 1 item                                                                         
test/test_cloud.py [2024-01-23 16:12:33,345] [ingestion] [setup] [DEBUG] tiledb.cloud=0.11.2, tiledb=(0, 25, 0), libtiledb=(2, 19, 0)
[2024-01-23 16:12:33,345] [ingestion] [ingest] [DEBUG] Ingesting Vectors into 'tiledb://paris-morgan/s3://tiledb-paris/tiledb/groups/zzz_unittest_vector_search_YKeuOcOXpO/test_ivf_flat_random_sampling_array'
[2024-01-23 16:12:37,532] [ingestion] [ingest] [DEBUG] Input dataset size 10000
[2024-01-23 16:12:37,532] [ingestion] [ingest] [DEBUG] Input dataset dimensions 128
[2024-01-23 16:12:37,532] [ingestion] [ingest] [DEBUG] Vector dimension type float32
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] Partitions 100
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] Training sample size 1234
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] Training source uri None and type None
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] Number of workers 10
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] input_vectors_per_work_item 20000000
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] input_vectors_work_items 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] input_vectors_work_tasks 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] input_vectors_work_items_per_worker 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] input_vectors_per_work_item_during_sampling 1000000
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] input_vectors_work_items_during_sampling 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] input_vectors_work_items_per_worker_during_sampling 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] table_partitions_per_work_item 200000
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] table_partitions_work_items 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] table_partitions_work_tasks 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] table_partitions_work_items_per_worker 1
[2024-01-23 16:12:37,533] [ingestion] [ingest] [DEBUG] Creating arrays
[2024-01-23 16:13:35,107] [ingestion] [create_arrays] [DEBUG] Creating temp ids array
[2024-01-23 16:13:35,107] [ingestion] [create_arrays] [DEBUG] ArraySchema(
  domain=Domain(*[
    Dim(name='rows', domain=(0, 2147483647), tile=125000, dtype='int32'),
  ]),
  attrs=[
    Attr(name='values', dtype='uint64', var=False, nullable=False, enum_label=None, filters=FilterList([ZstdFilter(level=-1), ])),
  ],
  cell_order='col-major',
  tile_order='col-major',
  sparse=False,
)
[2024-01-23 16:13:41,576] [ingestion] [create_arrays] [DEBUG] Creating temp parts array
[2024-01-23 16:13:41,576] [ingestion] [create_arrays] [DEBUG] ArraySchema(
  domain=Domain(*[
    Dim(name='rows', domain=(0, 127), tile=128, dtype='int32'),
    Dim(name='cols', domain=(0, 2147483647), tile=125000, dtype='int32'),
  ]),
  attrs=[
    Attr(name='values', dtype='float32', var=False, nullable=False, enum_label=None, filters=FilterList([ZstdFilter(level=-1), ])),
  ],
  cell_order='col-major',
  tile_order='col-major',
  sparse=False,
)
[2024-01-23 16:13:41,576] [ingestion] [create_arrays] [DEBUG] tiledb://paris-morgan/s3://tiledb-paris/tiledb/groups/zzz_unittest_vector_search_YKeuOcOXpO/test_ivf_flat_random_sampling_array/temp_data/shuffled_vectors
[2024-01-23 16:13:47,825] [ingestion] [create_arrays] [DEBUG] Creating part array tiledb://paris-morgan/s3://tiledb-paris/tiledb/groups/zzz_unittest_vector_search_YKeuOcOXpO/test_ivf_flat_random_sampling_array/temp_data/partition_indexes/0
[2024-01-23 16:13:47,825] [ingestion] [create_arrays] [DEBUG] ArraySchema(
  domain=Domain(*[
    Dim(name='rows', domain=(0, 100), tile=100, dtype='int32'),
  ]),
  attrs=[
    Attr(name='values', dtype='uint64', var=False, nullable=False, enum_label=None, filters=FilterList([ZstdFilter(level=-1), ])),
  ],
  cell_order='col-major',
  tile_order='col-major',
  sparse=False,
)
[2024-01-23 16:14:12,672] [ingestion] [ingest] [DEBUG] Creating ingestion graph
[2024-01-23 16:14:16,359] [ingestion] [create_array] [DEBUG] Creating input vectors array
[2024-01-23 16:14:16,359] [ingestion] [create_array] [DEBUG] ArraySchema(
  domain=Domain(*[
    Dim(name='rows', domain=(0, 127), tile=128, dtype='int32'),
    Dim(name='cols', domain=(0, 1233), tile=1234, dtype='int32'),
  ]),
  attrs=[
    Attr(name='values', dtype='float32', var=False, nullable=False, enum_label=None, filters=FilterList([ZstdFilter(level=-1), ])),
  ],
  cell_order='col-major',
  tile_order='col-major',
  sparse=False,
)
[ingestion@create_ingestion_dag] num_sampled 1234
[ingestion@create_ingestion_dag] training_source_uri tiledb://paris-morgan/s3://tiledb-paris/tiledb/groups/zzz_unittest_vector_search_YKeuOcOXpO/test_ivf_flat_random_sampling_array/training_input_vectors
[ingestion@create_ingestion_dag] uri tiledb://paris-morgan/s3://tiledb-paris/tiledb/groups/zzz_unittest_vector_search_YKeuOcOXpO/test_ivf_flat_random_sampling_array/training_input_vectors, uri_type TILEDB_ARRAY // training_source_uri tiledb://paris-morgan/s3://tiledb-paris/tiledb/groups/zzz_unittest_vector_search_YKeuOcOXpO/test_ivf_flat_random_sampling_array/training_input_vectors, training_source_type TILEDB_ARRAY
[ingestion@create_ingestion_dag] start 0, end 10000
[ingestion@create_ingestion_dag] start 0, end 10000
[ingestion@create_ingestion_dag] start 0, end 10000
[ingestion@create_ingestion_dag] start 0, end 10000
[ingestion@create_ingestion_dag] start 0, end 10000
[2024-01-23 16:14:27,663] [ingestion] [ingest] [DEBUG] Submitting ingestion graph
[2024-01-23 16:14:34,459] [ingestion] [ingest] [DEBUG] Submitted ingestion graph
```
</details>

The key part is `IndexError: index 1234 is out of bounds for axis 0 with size 1234`, i.e. we are trying to access the 1234th row of the training input vectors, but there are only 1234 rows.
    
With this fix we now don't crash, as we can see in these logs from this job: https://cloud.tiledb.com/activity/tasks/27e45a6e-6490-4a32-a2f9-55185f24c05f
```
[2024-01-24 12:37:40,491] [ingestion] [update_centroids] [DEBUG] Using 16 threads.
[ingestion@update_centroids] len(vectors) 1234 batch_size 78
[ingestion@update_centroids] i 0 start 0 end 78 / previous start 0 previous end 78
[2024-01-24 12:37:40,497] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Vectors computed: 0
[ingestion@update_centroids] i 78 start 78 end 156 / previous start 78 previous end 156
[2024-01-24 12:37:40,499] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 0
[ingestion@update_centroids] i 156 start 156 end 234 / previous start 156 previous end 234
[2024-01-24 12:37:40,505] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 1
[ingestion@update_centroids] i 234 start 234 end 312 / previous start 234 previous end 312
[2024-01-24 12:37:40,510] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 2
[ingestion@update_centroids] i 312 start 312 end 390 / previous start 312 previous end 390
[2024-01-24 12:37:40,536] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 3
[ingestion@update_centroids] i 390 start 390 end 468 / previous start 390 previous end 468
[2024-01-24 12:37:40,540] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 4
[ingestion@update_centroids] i 468 start 468 end 546 / previous start 468 previous end 546
[2024-01-24 12:37:40,544] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 5
[ingestion@update_centroids] i 546 start 546 end 624 / previous start 546 previous end 624
[2024-01-24 12:37:40,549] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 6
[ingestion@update_centroids] i 624 start 624 end 702 / previous start 624 previous end 702
[2024-01-24 12:37:40,553] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 7
[ingestion@update_centroids] i 702 start 702 end 780 / previous start 702 previous end 780
[2024-01-24 12:37:40,557] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 8
[ingestion@update_centroids] i 780 start 780 end 858 / previous start 780 previous end 858
[2024-01-24 12:37:40,561] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 9
[ingestion@update_centroids] i 858 start 858 end 936 / previous start 858 previous end 936
[2024-01-24 12:37:40,566] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 10
[ingestion@update_centroids] i 936 start 936 end 1014 / previous start 936 previous end 1014
[2024-01-24 12:37:40,570] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 11
[ingestion@update_centroids] i 1014 start 1014 end 1092 / previous start 1014 previous end 1092
[2024-01-24 12:37:40,574] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 12
[2024-01-24 12:37:40,578] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 13
[ingestion@update_centroids] i 1092 start 1092 end 1170 / previous start 1092 previous end 1170
[ingestion@update_centroids] i 1170 start 1170 end 1234 / previous start 1170 previous end 1248
[2024-01-24 12:37:40,600] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 14
[2024-01-24 12:37:40,603] [ingestion] [generate_new_centroid_per_thread] [DEBUG] Finished thread: 15
[2024-01-24 12:37:40,606] [ingestion] [update_centroids] [DEBUG] Finished all threads, aggregating partial results.
```
Here `start / end` are the new values and and `previous start / end` are the old values. We don't run over now: `[ingestion@update_centroids] i 1170 start 1170 end 1234 / previous start 1170 previous end 1248`.